test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Common/CustomHttpClientRequestMessage.cs

```diff
@@ -0,0 +1,127 @@
+    /// </summary>
+    public class CustomHttpClientRequestMessage : IODataRequestMessageAsync, IServiceCollectionProvider
+    {
+        private readonly HttpRequestMessage _request;
+        private readonly HttpClient _httpClient;
+        private HttpContent _content;
+        private MemoryStream _memoryStream;
+
+        public CustomHttpClientRequestMessage(Uri uri, HttpClient httpClient)
+        {
+            _httpClient = httpClient;
+            _memoryStream = new MemoryStream();
+
+            _request = new HttpRequestMessage
+            {
+                RequestUri = uri
+            };
+        }
+
+        public string? GetHeader(string headerName)
+        {
+            return _request.Headers.TryGetValues(headerName, out var values) ||
+                (_content?.Headers.TryGetValues(headerName, out values) ?? false)
+                ? string.Join(",", values)
+                : null;
+        }
+
+        public async Task<Stream> GetStreamAsync()
+        {
+            if (_memoryStream == null)
+            {
+                _memoryStream = new MemoryStream();
+                _content = new StreamContent(_memoryStream);
+                _request.Content = _content;
+            }
+            else
+            {
+                _memoryStream.Position = _memoryStream.Length;
+            }
+
+            return await Task.FromResult<Stream>(_memoryStream);
+        }
+
+        public IEnumerable<KeyValuePair<string, string>> Headers
+        {
+            get
+            {
+                if (_content != null)
+                {
+                    foreach (var contentHeader in _content.Headers)
+                    {
+                        yield return new KeyValuePair<string, string>(contentHeader.Key, string.Join(",", contentHeader.Value));
+                    }
+                }
+            }
+        }
+
+        public void SetHeader(string headerName, string headerValue)
+        {
+            if (!_request.Headers.TryAddWithoutValidation(headerName, headerValue))
+            {
+                if (_content == null)
+                {
+                    _content = new StreamContent(_memoryStream);
+                    _request.Content = _content;
+                }
+                _content.Headers.TryAddWithoutValidation(headerName, headerValue);
+            }
+        }
+
+        public async Task<IODataResponseMessageAsync> GetResponseAsync()
+        {
+            HttpResponseMessage response;
+
+            try
+            {
+                response = await _httpClient.SendAsync(_request);
+            }
+            catch (HttpRequestException ex)
+            {
+                throw new Exception($"Request to {_request.RequestUri} failed: {ex.Message}", ex);
+            }
+
+            return new HttpClientResponseMessage(response)
+            {
+                ServiceProvider = ServiceProvider
+            };
+        }
+
+        public Stream GetStream()

```

Comment:
How come the implementations for `GetStream` and `GetStreamAsync` are so different yet they're supposed to complement each other?
Is there a way you can avoid calling an async method from `GetStream` method?

Reply:
Yes. I'll update that.